lib/libutils/ext/timingsafe_memcmp.c

```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: ISC
+/*	$OpenBSD: timingsafe_memcmp.c,v 1.2 2015/08/31 02:53:57 guenther Exp $	*/
+/*
+ * Copyright (c) 2014 Google Inc.
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <limits.h>
+#include <string_ext.h>
+
+int
+timingsafe_memcmp(const void *b1, const void *b2, size_t len)
+{
+        const unsigned char *p1 = b1, *p2 = b2;
+        size_t i;
+        int res = 0, done = 0;
+
+        for (i = 0; i < len; i++) {
+                /* lt is -1 if p1[i] < p2[i]; else 0. */
+                int lt = (p1[i] - p2[i]) >> CHAR_BIT;
```

Comment:
Well `E1` has an unsigned type, no? (`const unsigned char - const unsigned char`).

Reply:
As I read the standard `(p1[i] - p2[i])` is promoted to the type `int`.
If it would have remained an `unsigned char` the result would always be `0` due to the shift.

Reply:
Someone other than me has obviously noticed this also. Here's a better version:
 http://bk.ntp.org/ntp-stable/lib/isc/tsmemcmp.c
```
%sha1sum tsmemcmp.c
7e4f6a83cfe2c4956bfa4a81e47b9bec906ae128  tsmemcmp.c

Reply:
That's quite helpful :+1: I will import this version instead.